chain/chain/src/chain.rs

```diff
@@ -5099,21 +5101,36 @@ impl<'a> ChainUpdate<'a> {
                     .get_split_shard_uids(shard_uid.shard_id())
                     .unwrap_or_else(|| panic!("invalid shard layout {:?}", next_epoch_shard_layout))
                     .len() as NumShards;
+
                 let total_gas_used = chunk_extra.gas_used();
                 let total_balance_burnt = chunk_extra.balance_burnt();
-                let gas_res = total_gas_used % num_split_shards;
+
+                let mut gas_res = total_gas_used % num_split_shards;
                 let gas_split = total_gas_used / num_split_shards;
-                let balance_res = (total_balance_burnt % num_split_shards as u128) as NumShards;
+
+                let mut balance_res = (total_balance_burnt % num_split_shards as u128) as NumShards;
                 let balance_split = total_balance_burnt / (num_split_shards as u128);
                 let gas_limit = chunk_extra.gas_limit();
                 let outcome_root = *chunk_extra.outcome_root();
 
                 let mut sum_gas_used = 0;
                 let mut sum_balance_burnt = 0;
+
+                // The gas and balance distribution assumes that we have a result for every split shard.
+                // TODO(resharding) make sure that is the case.
+                assert_eq!(num_split_shards, results.len() as u64);
+
                 for result in results {
-                    let shard_id = result.shard_uid.shard_id();
-                    let gas_burnt = gas_split + if shard_id < gas_res { 1 } else { 0 };
-                    let balance_burnt = balance_split + if shard_id < balance_res { 1 } else { 0 };
+                    let mut gas_burnt = gas_split;
+                    if gas_res > 0 {
+                        gas_res -= 1;
+                        gas_burnt += 1;
+                    }
```

Comment:
If we only need gas_burnt to be mut for this block, may I suggest something like the following to make it even easier for the reader.
```
let gas_burnt = {
                    let mut gas_burnt = gas_split;
                    if gas_res > 0 {
                        gas_res -= 1;
                        gas_burnt += 1;
                    }
                    gas_burnt
};
```